sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py

```diff
@@ -84,19 +86,60 @@ class CachedNodeResolver(object):
     def __init__(
         self,
         resolver,
-        subscription_id: str,
-        resource_group_name: str,
-        workspace_name: str,
-        registry_name: str,
+        subscription_id: Optional[str],
+        resource_group_name: Optional[str],
+        workspace_name: Optional[str],
+        registry_name: Optional[str],
     ):
         self._resolver = resolver
         self._cache: Dict[str, _CacheContent] = {}
         self._nodes_to_resolve: List[BaseNode] = []
 
+        self._client_hash = self._get_client_hash(
+            subscription_id, resource_group_name, workspace_name, registry_name
+        )
+
+    @staticmethod
+    def _get_client_hash(
+        subscription_id: Optional[str],
+        resource_group_name: Optional[str],
+        workspace_name: Optional[str],
+        registry_name: Optional[str],
+    ) -> str:
+        """Get a hash for used client.
+        Works for both workspace client and registry client.
+        """
         object_hash = hashlib.sha256()
         for s in [subscription_id, resource_group_name, workspace_name, registry_name]:
             object_hash.update(str(s).encode("utf-8"))
-        self._workspace_hash = object_hash.hexdigest()
+        return object_hash.hexdigest()
+
+    @staticmethod
+    def _get_component_registration_max_workers():
+        """Get the max workers for component registration.
+
+        Before Python 3.8, the default max_worker is the number of processors multiplied by 5.
+        It may send a large number of the uploading snapshot requests that will occur remote refuses requests.
+        In order to avoid retrying the upload requests, max_worker will use the default value in Python 3.8,
+        min(32, os.cpu_count + 4).
+
+        1 risk is that, asset_utils will create a new thread pool to upload files in subprocesses, which may cause
+        the number of threads exceed the max_worker.
+        """
+        default_max_workers = min(32, (os.cpu_count() or 1) + 4)
```

Comment:
I think or is used with booleans?  

what will the `or 1` do?

Reply:
`x or 1` means `1 if x is None else x`